crates/hir-def/src/body/lower.rs

```diff
@@ -302,16 +302,29 @@ impl ExprCollector<'_> {
                 self.alloc_expr(Expr::For { iterable, pat, body, label }, syntax_ptr)
             }
             ast::Expr::CallExpr(e) => {
-                let callee = self.collect_expr_opt(e.expr());
-                let args = if let Some(arg_list) = e.arg_list() {
-                    arg_list.args().filter_map(|e| self.maybe_collect_expr(e)).collect()
-                } else {
-                    Box::default()
+                let is_rustc_box = {
+                    let attrs = e.attrs();
+                    attrs.filter_map(|x| x.as_simple_atom()).any(|x| x == "rustc_box")
                 };
-                self.alloc_expr(
-                    Expr::Call { callee, args, is_assignee_expr: self.is_lowering_assignee_expr },
-                    syntax_ptr,
-                )
+                if is_rustc_box {
+                    let expr = self.collect_expr_opt(e.arg_list().and_then(|x| x.args().next()));
+                    self.alloc_expr(Expr::Box { expr }, syntax_ptr)
```

Comment:
Can we move this to mir lowering? I want to get rid of `Expr::Box` eventually since box expressions have been removed from the surface syntax.

Reply:
IIRC there was a problem in using attributes in hir, and the `#[rustc_box]` was only available in ast. The rustc also does this (for other reasons, I don't remember them currently so don't know how applicable they are to us) and they want to get rid of box in hir together with `#[rustc_box]` some day. Maybe we can keep it until that day, and only remove the `box` syntax from ast?

Reply:
Oh, I forgot we haven't implemented attributes on expressions in hir. That's pretty unfortunate but yeah we'd have to keep this way. cc #14504.